include/unifex/unstoppable_baton.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+
+#include <atomic>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _ub {
+
+struct _op_base;
+
+template <typename Receiver>
+struct _operation {
+  struct type;
+};
+
+template <typename Receiver>
+using operation = typename _operation<Receiver>::type;
+
+struct unstoppable_baton;
+
+struct _sender {
+  template <template <class...> class Variant,
+            template <class...> class Tuple>
+  using value_types = Variant<Tuple<>>;
+
+  template <template <class...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+
+  static constexpr bool sends_done = false;
+
+  explicit _sender(unstoppable_baton& baton) noexcept
+    : baton_(&baton) {}
+
+  template (typename Receiver)
+    (requires receiver_of<Receiver>)
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) const noexcept(
+      std::is_nothrow_constructible_v<std::decay_t<Receiver>, Receiver&&>) {
+    return operation<std::decay_t<Receiver>>{*baton_, (Receiver&&)r};
+  }
+
+ private:
+  unstoppable_baton* baton_;
+};
+
+struct unstoppable_baton {
+  unstoppable_baton() noexcept
+    : unstoppable_baton(false) {}
+
+  explicit unstoppable_baton(bool startSignalled) noexcept
+    : state_(startSignalled ? this : nullptr) {}
+
+  void post() noexcept;
+
+  bool ready() const noexcept {
+    return state_.load(std::memory_order_acquire) ==
+        static_cast<const void*>(this);
+  }
+
+  void reset() noexcept {
+    // transition from signalled (i.e. state_ == this) to not-signalled
+    // (i.e. state_ == nullptr).
+    void* oldState = this;
+
+    // We can ignore the the result.  We're using _strong so it won't fail
+    // spuriously; if it fails, it means it wasn't previously in the signalled
+    // state so resetting is a no-op.
+    (void)state_.compare_exchange_strong(
+        oldState, nullptr, std::memory_order_acq_rel, std::memory_order_relaxed);
```

Comment:
From the perspective of the caller, there's no indication that the operation either succeeded or failed so the caller can't make any assumptions about whether this method synchronizes or not -- thus making the asymmetry here confusing. Also, the release and acquire semantics don't seem to be needed because the new value (`nullptr`) doesn't have any dependent data to synchronize.

Reply:
The use case this was trying to satisfy was implementing a dirty-flag where one or more producers might add work to be done and then call post() on the baton to wake up the consumer, then the consumer would wake-up and immediately call reset() to clear the dirty-flag before processing changes so that it can be sure that it won't miss any subsequent wake-ups that might come up while processing the previous changes.

```
unstoppable_baton baton;
T state; // assume internally synchronised

void produce() {
  state.enqueue(item);  // 'release' semantics - synchronises with 'dequeue_all()'
  baton.post();
}

task consume() {
  while (true) {
    co_await baton.wait();
    baton.reset();
    auto items = state.dequeue_all(); // 'acquire' semantics here.
    process(items);
  }
}
```

Now let's say we have T1, T2 both calling produce() and T3 calling consume().
We want to make sure that we don't have any missed wake-ups of consume().

Consider what could happen if reset() has relaxed memory order.
I was worried that the write inside reset() might be reordered to occur after the state.dequeue_all() call (since they are occurring on unrelated objects), leaving the following program order.
```
T1: state.enqueue(item)
T1: baton.post()
T3: baton.wait() - sees already signalled
T3: state.dequeue_all() - sees item from T1
T2: state.enqueue(item)
T2: baton.post() - sees already signalled, does nothing
T3: baton.reset() - clears flag set by T2 from T
T3: ... processes item from T1 and continues around loop
T3: baton.wait() - blocks here even though there are unprocessed items in the queue
```

The intention behind making the reset() operation 'acquire' was to ensure that the subequent dequeue_all() operation is not reordered before the reset() from the perspective of T2. But it's possible this logic is flawed.

We need to ensure that either:
- T2's enqueue() appears in TMO of state before T3's dequeue() - i.e. that T3 will process T2's item
- T3's reset() appears in TMO of baton before T3's post() - i.e. that T2's subsequent post() will signal the baton so that T3's subsequent wait() will see it

It's possible that we may actually need to strengthen the memory order here to seq_cst to satisfy this, although I'm not entirely sure...

Alternatively, this use-case might be better served by a different synchronisation primitive like an `auto_reset_event`.

Reply:
I think that the use case that you provided makes sense for the acquire semantics here. From my perspective this is a good argument to strengthen the memory ordering for `reset`. 

To expand on your example, if there are two consumers each resetting the baton, then only one of them will have acquire semantics -- any other consumer would have relaxed semantics which means they wouldn't synchronize and could receive stale results.

Reply:
> From my perspective this is a good argument to strengthen the memory ordering for `reset`

Do you mean strengthening to seq_cst? Or do you think acq_rel is strong enough?

> if there are two consumers each resetting the baton

This pattern really only works correctly for a single consumer processing the dirty flag/event and resetting it.

For multiple consumers an `async_auto_reset_event` would be more appropriate.

Reply:
> Do you mean strengthening to seq_cst? Or do you think acq_rel is strong enough?

`acq_rel` should be strong enough here

> This pattern really only works correctly for a single consumer processing the dirty flag/event and resetting it.
> For multiple consumers an async_auto_reset_event would be more appropriate.

I might be overthinking this then -- I'm not sure of the intended use case of this class so if the single consumer case makes sense to optimize for then let's go with `acq_rel` semantics.

Reply:
Addressed in 6b39655.